File: commonroad_reach_semantic/data_structure/model_checking/finite_automaton.py
import functools
from collections import defaultdict
from typing import Iterator, List, Dict, Iterable, Tuple, FrozenSet

import buddy
import spot

import commonroad_reach_semantic.utility.spot as util_spot


class FiniteAutomaton:
    """Represents a finite automaton on words over the powerset of propositions."""

    _spot_automaton: spot.twa_graph
    _bdict: spot.bdd_dict

    def __init__(self, ltlf_formulas: List[str], mode: int = 0) -> None:
        """Construct the finite automaton accepting those words that make the given LTLf formula true

        :param ltlf_formulas: The LTLf formulas to translate.
        :param mode: How to combine the formulas into a single automaton:
            0 = decide automatically
            1 = translate individually and then compute the product automaton (faster)
            2 = form conjunction of the formulas and translate the result (probably smaller automaton)
        """
        if mode == 0:
            # We always use the product automaton for now
            mode = 1

        spot_formulas = [spot.from_ltlf(f) for f in ltlf_formulas]
        if mode == 1:
            buechi_automata = [self._translate_ltlf_to_buechi(f) for f in spot_formulas]
            product_automaton = functools.reduce(spot.product, buechi_automata)
            self._spot_automaton = spot.to_finite(product_automaton)
        elif mode == 2:
            conjunction = self._formula_conjunction(spot_formulas)
            self._spot_automaton = spot.to_finite(self._translate_ltlf_to_buechi(conjunction))
        else:
            raise ValueError("Invalid mode")

        self._bdict = self._spot_automaton.get_dict()

    @property
    def initial_state(self) -> int:
        """The number of the initial state."""
        return self._spot_automaton.get_init_state_number()

    def transitions_from(self, state: int) -> Iterator[Tuple[List[FrozenSet[Tuple[str, bool]]], int]]:
        """Iterate over all transitions outgoing from the given state."""
        for edge in self._spot_automaton.out(state):
            yield self._edge_condition_to_minterms(edge.cond), edge.dst

    def combined_transitions_from(self, states: Iterable[int]) -> Iterator[Tuple[FrozenSet[Tuple[str, bool]], int]]:
        """Iterate over all transitions outgoing from the given states.

        Tries to minimize the minterms by combining the conditions of the outgoing edges leading to the same destination.
        """
        dst_state_to_conditions: Dict[int, List[buddy.bdd]] = defaultdict(list)
        for state in frozenset(states):
            for edge in self._spot_automaton.out(state):
                dst_state_to_conditions[edge.dst].append(edge.cond)
        for dst_state, conditions in dst_state_to_conditions.items():
            for minterm in self._edge_condition_to_minterms(functools.reduce(buddy.bdd_or, conditions)):
                yield frozenset(minterm), dst_state

    def is_accepting_state(self, state: int) -> bool:
        """Check whether the given state is an accepting state.

        :returns: True if and only if the state is accepting.
        """
        return self._spot_automaton.state_is_accepting(state)

    def _edge_condition_to_minterms(self, cond: buddy.bdd) -> List[FrozenSet[Tuple[str, bool]]]:
        """Convert a condition on an automaton edge given as a BDD into a list of minterms.

        The condition is true iff at least one minterm is satisfied.
        A minterm is a list of possibly negated atomic propositions.
        It is satisfied iff all its atomic propositions hold.
        """
        # will be in DNF --> bbd_to_formula computes an irredundant sum of products
        # https://spot.lre.epita.fr/doxygen/namespacespot.html#aba9b9efe994006c29a6d77da94897df8
        formula_dnf = spot.bdd_to_formula(cond, self._bdict)
        return [frozenset(minterm) for minterm in util_spot.extract_minterms_from_dnf(formula_dnf)]

    @staticmethod
    def _formula_conjunction(formulas: List[spot.formula]) -> spot.formula:
        """Conjunction the given LTLf formulas."""
        return spot.formula.And(formulas)

    @staticmethod
    def _translate_ltlf_to_buechi(ltlf_formula: spot.formula) -> spot.twa_graph:
        """Translate the given LTLf formula into a Buechi automaton."""
        # disable simulation based reductions to speed up translation
        # see https://spot.lre.epita.fr/man/spot-x.7.html
        return ltlf_formula.translate(xargs="simul=0")

File: cpp/src/data_structure/model_checking/finite_automaton.cpp
#include "reach_semantic/data_structure/model_checking/finite_automaton.hpp"
#include "reach_semantic/utility/spot.hpp"
#include <spot/misc/optionmap.hh>
#include <spot/twa/formula2bdd.hh>
#include <spot/twaalgos/product.hh>
#include <utility>

using namespace semantic_reach;

FiniteAutomaton::FiniteAutomaton(const std::vector<std::string> &ltlf_formulas, int mode) {
    if (mode == 0) {
        // We always use the product automaton for now
        mode = 1;
    }

    std::vector<spot::formula> spot_formulas;
    spot_formulas.reserve(ltlf_formulas.size());
    std::transform(ltlf_formulas.begin(), ltlf_formulas.end(), std::back_inserter(spot_formulas),
                   [](const std::string &formula) {
                       spot::parsed_formula parsed_formula = spot::parse_infix_psl(formula);
                       if (parsed_formula.format_errors(std::cerr)) {
                           throw std::runtime_error("Error while parsing LTLf formula.");
                       }
                       return spot::from_ltlf(parsed_formula.f);
                   });

    auto trans{_get_translator()};

    switch (mode) {
    case 1: {
        auto true_automaton = trans.run(spot::formula::tt());
        auto product_automaton =
            std::accumulate(spot_formulas.begin(), spot_formulas.end(), true_automaton,
                            [&trans](const spot::twa_graph_ptr &acc, const spot::formula &formula) {
                                return spot::product(acc, trans.run(formula));
                            });
        _spot_automaton = spot::to_finite(product_automaton);
        break;
    }
    case 2: {
        auto conjunction = spot::formula::And(std::move(spot_formulas));
        _spot_automaton = spot::to_finite(trans.run(conjunction));
        break;
    }
    default:
        throw std::runtime_error("Invalid mode for combining LTLf formulas.");
    }

    _bdict = _spot_automaton->get_dict();
}

State FiniteAutomaton::initial_state() { return _spot_automaton->get_init_state_number(); }

std::vector<std::pair<Minterm, State>> FiniteAutomaton::transitions_from(const std::set<State> &states) {
    std::map<State, std::vector<bdd>> map_state_to_conditions{};
    for (const auto &state : states) {
        for (auto &edge : _spot_automaton->out(state)) {
            map_state_to_conditions[edge.dst].emplace_back(edge.cond);
        }
    }

    std::vector<std::pair<Minterm, State>> result{};
    for (const auto &[dst_state, conditions] : map_state_to_conditions) {
        bdd combined{bdd_false()};
        for (const auto &condition : conditions) {
            combined = bdd_or(combined, condition);
        }
        auto minterms_to_dst = _edge_condition_to_minterms(std::move(combined));
        for (auto &&minterm : minterms_to_dst) {
            result.emplace_back(std::move(minterm), dst_state);
        }
    }
    return result;
}

std::map<Minterm, std::set<State>> FiniteAutomaton::multi_transitions_from(const std::set<State> &states) {
    if (_multi_transitions_cache.count(states) != 0) {
        return _multi_transitions_cache.at(states);
    }

    std::map<Minterm, std::set<State>> minterm_to_dst_states{};
    for (const auto &[minterm, dst_state] : transitions_from(states)) {
        minterm_to_dst_states[minterm].emplace(dst_state);
    }
    _multi_transitions_cache.insert({states, minterm_to_dst_states});
    return minterm_to_dst_states;
}

bool FiniteAutomaton::is_accepting_state(State state) { return _spot_automaton->state_is_accepting(state); }

std::vector<Minterm> FiniteAutomaton::_edge_condition_to_minterms(bdd cond) {
    // will be in DNF --> bbd_to_formula computes an irredundant sum of products
    // https://spot.lre.epita.fr/doxygen/namespacespot.html#aba9b9efe994006c29a6d77da94897df8
    auto formula_dnf = spot::bdd_to_formula(std::move(cond), _bdict);
    return util_spot::extract_minterms_from_dnf(formula_dnf);
}

spot::translator FiniteAutomaton::_get_translator() {
    spot::option_map options{};
    // disable simulation based reductions to speed up translation
    // see https://spot.lre.epita.fr/man/spot-x.7.html
    options.set("simul", 0);
    spot::translator trans{&options};
    trans.set_type(spot::postprocessor::Buchi);
    trans.set_pref(spot::postprocessor::SBAcc | spot::postprocessor::Small);
    return trans;
}
